Add support for aliases

michaelkuhn opened the pull request:
Aliases can be used to define new commands. For instance, `sp: spec -I` will define a new command `sp` that will execute `spec` with the `-I` argument. Aliases cannot override existing commands.

This is still WIP and is missing some things, including schema adaptions and porting the old aliases to the new system. It already works in general. For convenience, there is an example alias `sp` (which I will remove for the final version). I would just like to get some feedback on whether this is something of interest.

@adamjstewart @tgamblin @alalazo

Closes #2705

adamjstewart commented:
Also see #2705

michaelkuhn commented:
I have read that PR and think the main issue back then was that standard commands could be overridden. This PR does not allow that (like Git).

michaelkuhn commented:
Looks like I'm hitting some strange test failures due to using `spack.config.get` in the main function. I have opened #17262 about it.

tgamblin commented:
I like it!  Go for it!

michaelkuhn commented:
This is ready from my side. If you have any more comments, please let me know.

michaelkuhn commented:
@tgamblin @adamjstewart ping

adamjstewart commented:
I don't usually merge core changes, but I do agree that this is a nice feature to add. Maybe @becker33 or @scheibelp can review?

michaelkuhn commented:
@tgamblin @becker33 @scheibelp ping

I have fixed a conflict, so this should be ready again.

michaelkuhn committed:
Add support for aliases

Aliases can be used to define new commands. For instance, `sp: spec -I`
will define a new command `sp` that will execute `spec` with the `-I`
argument. Aliases cannot override existing commands.

michaelkuhn commented:
@tgamblin @becker33 @scheibelp ping ping

michaelkuhn commented:
This has been finished for some time from my point of view, I would *really* appreciate a review.

adamjstewart commented:
How does this work for subsubcommands, like `spack compiler find`? Does it only support aliases of subcommands? Do we want support for subsubcommands? Is that even possible?

michaelkuhn commented:
It currently only supports creating aliases on a subcommand level. You can, however, map them so subsubcommands. For instance, the following works: `foo: compiler list`

Edit: Looking at the code again, it might be possible to support them on a subsubcommand level. I'm not sure whether that's a good idea, though.

becker33 commented on lib/spack/spack/main.py:
```diff
@@ -313,7 +309,10 @@ def add_command(self, cmd_name):
+        alias_list = []
+        aliases = spack.config.get('config:aliases')
+        if aliases:
+            alias_list = [k for k, v in aliases.items() if v == cmd_name]
```
This needs to check whether the first word of v is cmd_name, not whether they are entirely equal.

michaelkuhn replied:
Good catch, thanks!

becker33 commented on lib/spack/spack/main.py:
```diff
@@ -752,12 +751,22 @@ def main(argv=None):
-        cmd_name = aliases.get(cmd_name, cmd_name)
+
+        if cmd_name not in spack.cmd.all_commands():
+            alias = spack.config.get('config:aliases')
+
+            if alias:
+                alias = alias.get(cmd_name)
```
```suggestion
            aliases = spack.config.get('config:aliases')

            if aliases:
                alias = alias.get(cmd_name)
```

scheibelp commented:
I can look at this by June 14th (one week from now)

becker33 commented on lib/spack/spack/main.py:
```diff
@@ -752,12 +751,22 @@ def main(argv=None):
 
         # Re-parse with the proper sub-parser added.
-        args, unknown = parser.parse_known_args()
+        args, unknown = parser.parse_known_args(args.command)
```
I'm almost surprised this works as cleanly as it does...

adamjstewart commented:
Another feature I've thought about (similar to aliases) is being able to abbreviate arguments, like https://docs.python.org/3/library/argparse.html#argument-abbreviations-prefix-matching.

Spack doesn't yet support this for subcommands but it might be useful to add. For example, if I add an alias in my bashrc from `spack` to `s`, I could type `s ins` instead of `spack install` because there is only one subcommand that starts with `ins`.

Not sure how others feel about this. I don't want to delay this PR by adding new features now, but something to think about if it's easy.

michaelkuhn commented:
I have just pushed an updated version that should address all of @becker33's comments.

Moreover, I have added another commit adding @adamjstewart's prefix matching proposal. Since it is a separate commit (and just one else branch), we can easily get rid of it if we do not want to include this functionality.

tgamblin commented on share/spack/spack-completion.bash:
```diff
@@ -333,7 +333,7 @@ _spack() {
         SPACK_COMPREPLY="-h --help -H --all-help --color -c --config -C --config-scope -d --debug --timestamp --pdb -e --env -D --env-dir -E --no-env --use-env-repo -k --insecure -l --enable-locks -L --disable-locks -m --mock -p --profile --sorted-profile --lines -v --verbose --stacktrace -V --version --print-shell-vars"
     else
-        SPACK_COMPREPLY="activate add analyze arch audit blame build-env buildcache cd checksum ci clean clone commands compiler compilers concretize config containerize create deactivate debug dependencies dependents deprecate dev-build develop docs edit env extensions external fetch find flake8 gc gpg graph help info install license list load location log-parse maintainers mark mirror module mo ...
+        SPACK_COMPREPLY="activate add analyze arch audit blame build-env buildcache cd checksum ci clean clone commands compiler compilers concretize config containerize containerise create deactivate debug dependencies dependents deprecate dev-build develop docs edit env extensions external fetch find flake8 gc gpg graph help info install license list load location log-parse maintainers mark mir ...
```
Can you modify the completion so that aliases are not included here?  This will force (as @haampie notes) the user to disambiguate not only among valid commands, but also among American and British English, which seems... annoying.

michaelkuhn replied:
Should we get rid of all aliases in the completions (for example, `compiler add` is an alias for `compiler find`) or only the ones defined in `config.yaml`?

michaelkuhn commented:
I have resolved the conflicts and rebased this to current develop. I have also removed the controversial `containerise` alias for now. If we want to keep it, I can give it a try in a follow-up PR.

michaelkuhn commented:
@alalazo @tgamblin yearly ping :wink:

I have fixed all conflicts and rebased to current develop.

psakievich commented on lib/spack/spack/main.py:
```diff
@@ -962,7 +963,28 @@ def _main(argv=None):
+            cmd_name = alias_parts[0]
+            args.command = alias_parts + args.command[1:]
+        else:
+            # If cmd_name is not an alias but also not a proper command,
```
This seems new to me, and missing from the docs.  Am I understanding this that is like an under the hood autocomplete?

`spack ins zlib` would automatically get mapped to `spack install zlib`

Or is it integral to how the aliases work and I'm just missing it?

michaelkuhn replied:
This part was motivated/requested by @adamjstewart above. It's a separate commit and can be dropped easily if we don't want to include that functionality.

psakievich replied:
I have mixed feelings about it.  I like it and I don't.  Encourages shortcuts which I love, could be confusing if you don't understand the metrics being used. 

My vote would be drop it so people have to explicitly define their aliases, but if we keep it I think it needs to be in the docs so we have an easy way to reference/explain the behavior when there are questions in the future.

adamjstewart replied:
My vote is to keep it obviously 😄 

It's not confusing if you don't know it exists and don't use it.

psakievich replied:
Sure. I just think the case to _not know it exists and accidentally use it_ is high.  We're psuedo supporting typos.  If there is a strong desire to keep then I just want us to be explicit that it is there in our documentation. I mean, tbh, I would probably use it now that I know it is there. I just don't look forward to explaining it to a bunch of people who ask me "why does spack do this?" 

Enough people in my sphere of influence already consider spack to be the great and mysterious Oz who does things the way it wants and for reasons beyond their comprehension.  I don't want to accidentally reinforce that sterotype for power user and core dev conveniences. 😅

psakievich replied:
Actually, I would fully support this if it were enabled by another config flag. I think a `power_user_mode` config option would be a nice umbrella flag for these sorts of things.  maybe separate PR though?

psakievich replied:
Any thoughts @becker33 or @tgamblin? seems like a discussion you two would want to weigh in on.

adamjstewart replied:
I would be fine with a power_user_mode flag but personally don't think it's necessary. Curious what others think though.

michaelkuhn replied:
I also don't think we need a power user mode. Other tools like `ip` also have this behavior. However, as I've mentioned, it's a separate commit and I would rather remove it before adding a power user mode or holding this PR up any longer.

adamjstewart replied:
@tgamblin can we sneak this into v0.21?

psakievich replied:
@becker33 @tgamblin??? I would love that @adamjstewart. Not sure if there is the bandwidth. Let's see. 🤞

adamjstewart replied:
Also, tab completion somewhat negates the need for the feature I requested so don't worry about me.

michaelkuhn committed:
Add alias test

The test makes sure that aliases cannot override Spack commands. This
also fixes a small bug in SpackCommand and delays the `add_command` call
until it is actually necessary; otherwise, it will complain about
aliases it does not know about.

michaelkuhn commented:
I've pushed a new version that (based on @psakievich's suggestion) adds a test for making sure that aliases cannot override commands.

tgamblin commented:
@haampie #40890

tgamblin commented:
@haampie: #40890.

tgamblin commented on lib/spack/spack/main.py:
```diff
@@ -870,6 +871,41 @@ def restore_macos_dyld_vars():
+            cmd_name = alias_parts[0]
+            cmd = alias_parts + cmd[1:]
+        else:
+            # If cmd_name is not an alias but also not a proper command,
+            # check whether it is a prefix of a proper command.
+            prefix = [c for c in spack.cmd.all_commands() if c.startswith(cmd_name)]
+
+            # We will use the found command only if there is exactly one
+            # match.
+            if len(prefix) == 1:
+                cmd_name = prefix[0]
+                cmd = [cmd_name] + cmd[1:]
```
I'd be ok with a better "did you mean?" error message for this, but I don't think we should automatically run a command the user didn't actually type, even if the prefix is unique.  We have completion for that.

Rationale: I think this can lead to sloppy scripting. It allows people to write, e.g., `spack inf hdf5`in a script and not get an error, which may one day become ambiguous. We should fail fast if you're not using valid commands.

michaelkuhn replied:
I have removed that part for now. I can give a separate "did you mean" functionality a try in a separate PR.

tgamblin replied:
Yeah let’s save “did you mean” for a separate PR, as I think people will have opinions.

tgamblin reviewed (changes requested):
I'm ok with this going in modulo the request below. Thoughts?

The error handling on this is a little loose, but that might be ok. For example, you can write:

```yaml
config:
    aliases:
        foo bar: info hdf5
```

and it shows up here:

```console
> spack -h
[...]

query packages:
  list                  list and search available packages
  info (foo bar)        get detailed information on a particular package

[...]
```

But you can't get it to run.  Not sure what to do with that, but it seems harmless.  It might be nice if it complained about spaces in alias keys if they're not supported, though.

I *think* it's actually good that it's quiet and just ignores aliases that override real commands.  It shouldn't be noisy about that, as you may alias something that isn't *yet* a spack command, but becomes one, and I don't think you'd want scripts to start failing if you accidentally happen to have config that overrides a real command.

haampie commented:
@tgamblin any chance to fix this issue still

```
$ spack -e . conc<tab>
...
$ spack -e . concreti
```

introduced in #39499 and not fixed by #39887. This PR doesn't seem to change it (didn't review it).

michaelkuhn commented:
> But you can't get it to run. Not sure what to do with that, but it seems harmless. It might be nice if it complained about spaces in alias keys if they're not supported, though.

That actually works, but apparently not through the shell wrapper:
```
$ ./bin/spack 'foo bar' 
CMakePackage:   hdf5
[...]
$ spack 'foo bar'                 
==> Error: foo is not a recognized Spack command or extension command; check with `spack commands`.
```
Do we want to make this work or should we just not support aliases with a space?

psakievich commented:
> I *think* it's actually good that it's quiet and just ignores aliases that override real commands.  It shouldn't be noisy about that, as you may alias something that isn't *yet* a spack command, but becomes one, and I don't think you'd want scripts to start failing if you accidentally happen to have config that overrides a real command.

In this case perhaps we want a warning?  If you had an alias that gets overridden by a new spack command **and** the new spack command changed the behavior that seems like it could be a subtle and annoying bug.

tgamblin commented:
@psakievich:

> In this case perhaps we want a warning?

It is probably helpful sometimes to spit out a warning but that’s as far as I’d go. Something like “alias ‘foo’ would run ‘bar’ which is not a spack command”, but again it could be annoying going back in time — if you alias a recent spack command then use an older version of spack.

@michaelkuhn? @adamjstewart?

tgamblin commented:
@michaelkuhn 
> Do we want to make this work or should we just not support aliases with a space?

I am having a hard time thinking of when I’d actually want a space… and we have no commands with spaces. Let’s disallow it. I can’t decide if it should just warn or if it should throw an error. Leaning towards error.

adamjstewart commented:
> @michaelkuhn
> 
> > Do we want to make this work or should we just not support aliases with a space?
> 
> I am having a hard time thinking of when I’d actually want a space… and we have no commands with spaces. Let’s disallow it. I can’t decide if it should just warn or if it should throw an error. Leaning towards error.

What about something like:
```yaml
config:
  aliases:
    comp a: compiler add
    comp f: compiler find
    ...
```

tgamblin commented:
@adamjstewart I think we should save that for later, for subcommand aliases.  Right now that won't be parsed the way I think you're expecting -- you'd have to quote it like `spack "comp a"`.

tgamblin commented:
So I guess given that how about a warning for aliases with spaces in them -- so that old versions of spack aren't too prickly if you end up with a configured subcommand alias for some future spack version.

michaelkuhn committed:
Sanity check aliases

Print a warning for aliases containing a space.

michaelkuhn commented:
I've pushed a new commit that will print a warning for aliases containing a space. And FYI, #40895 implements the "did you mean" functionality.

michaelkuhn committed:
Add warning for aliases attempting to override built-in commands

psakievich commented:
> @adamjstewart I think we should save that for later, for subcommand aliases.  Right now that won't be parsed the way I think you're expecting -- you'd have to quote it like `spack "comp a"`.

Or aliases of aliases... 😉

tgamblin committed:
fix documentation error, add types

adamjstewart commented:
3 years later. Glad we finally got this in!
